firedrake/tsfc_interface.py

```diff
@@ -131,15 +135,56 @@ def argument(self, o):
                                      "coefficient_map"])
 
 
-class TSFCKernel(DiskCached):
+class TSFCKernel(Cached):
 
     _cache = {}
-    if MPI.comm.rank == 0:
-        _cachedir = environ.get('FIREDRAKE_TSFC_KERNEL_CACHE_DIR',
-                                path.join(tempfile.gettempdir(),
-                                          'firedrake-tsfc-kernel-cache-uid%d' % getuid()))
-    else:
-        _cachedir = None
+
+    _cachedir = environ.get('FIREDRAKE_TSFC_KERNEL_CACHE_DIR',
+                            path.join(tempfile.gettempdir(),
+                                      'firedrake-tsfc-kernel-cache-uid%d' % getuid()))
+
+    @classmethod
+    def _cache_lookup(cls, key):
+        key, comm = key
+        return cls._cache.get(key) or cls._read_from_disk(key, comm)
+
+    @classmethod
+    def _read_from_disk(cls, key, comm):
+        if comm.rank == 0:
+            cache = cls._cachedir
+            filepath = os.path.join(cache, key)
+            val = None
+            if os.path.exists(filepath):
+                try:
+                    with gzip.open(filepath, 'rb') as f:
+                        val = f.read()
+                except zlib.error:
+                    pass
+
+            comm.bcast(val, root=0)
+        else:
+            val = comm.bcast(None, root=0)
+
+        if val is None:
+            raise KeyError("Object with key %s not found" % key)
+        val = cPickle.loads(val)
+        cls._cache[key] = val
+        return val
+
+    @classmethod
+    def _cache_store(cls, key, val):
+        key, comm = key
+        _ensure_cachedir(comm=comm)
+        if comm.rank == 0:
+            val._key = key
+            filepath = os.path.join(cls._cachedir, key)
+            tempfile = os.path.join(cls._cachedir, "%s_p%d.tmp" % (key, os.getpid()))
+            # No need for a barrier after this, since non root
+            # processes will never race on this file.
+            with gzip.open(tempfile, 'wb') as f:
+                cPickle.dump(val, f)
+            os.rename(tempfile, filepath)
+        comm.barrier()
```

Comment:
`TSFCKernel` was the only user of `DiskCached` so I move the implementation here (so that I can deal with communicators).